internal/querycoord/global_meta_broker.go

```diff
@@ -267,41 +269,121 @@ func (broker *globalMetaBroker) parseIndexInfo(ctx context.Context, segmentID Un
-	vecFieldIDs := funcutil.GetVecFieldIDs(schema)
-	if len(vecFieldIDs) != 1 {
-		err := fmt.Errorf("collection %d has multi vec field, num of vec fields = %d", collectionID, len(vecFieldIDs))
-		log.Error("get index info failed",
-			zap.Int64("collectionID", collectionID),
-			zap.Int64("segmentID", segmentID),
-			zap.Error(err))
-		return nil, err
-	}
-	indexInfo := &querypb.VecFieldIndexInfo{
-		FieldID: vecFieldIDs[0],
+// Better to let index params key appear in the file paths first.
+func (broker *globalMetaBroker) loadIndexExtraInfo(ctx context.Context, fieldPathInfo *indexpb.IndexFilePathInfo) (*extraIndexInfo, error) {
+	indexCodec := storage.NewIndexFileBinlogCodec()
+	for _, indexFilePath := range fieldPathInfo.IndexFilePaths {
+		// get index params when detecting indexParamPrefix
+		if path.Base(indexFilePath) == storage.IndexParamsKey {
+			content, err := broker.cm.MultiRead([]string{indexFilePath})
+			if err != nil {
+				return nil, err
+			}
+
+			if len(content) <= 0 {
+				return nil, fmt.Errorf("failed to read index file binlog, path: %s", indexFilePath)
+			}
+
+			indexPiece := content[0]
+			_, indexParams, indexName, _, err := indexCodec.Deserialize([]*storage.Blob{{Key: storage.IndexParamsKey, Value: []byte(indexPiece)}})
+			if err != nil {
+				return nil, err
+			}
+
+			return &extraIndexInfo{
+				indexName:   indexName,
+				indexParams: funcutil.Map2KeyValuePair(indexParams),
+			}, nil
+		}
 	}
-	// check the buildID of the segment's index whether exist on rootCoord
-	enableIndex, buildID, err := broker.getIndexBuildID(ctx, collectionID, segmentID)
+	return nil, errors.New("failed to load index extra info")
+}
+
+func (broker *globalMetaBroker) getFullIndexInfo(ctx context.Context, collectionID UniqueID, segmentID UniqueID) ([]*querypb.VecFieldIndexInfo, error) {
+	resp, err := broker.rootCoord.DescribeSegments(ctx, &rootcoordpb.DescribeSegmentsRequest{
+		Base: &commonpb.MsgBase{
+			MsgType: commonpb.MsgType_DescribeSegments,
+		},
+		CollectionID: collectionID,
+		SegmentIDs:   []UniqueID{segmentID},
+	})
 	if err != nil {
+		log.Error("failed to describe segments",
+			zap.Int64("collection", collectionID),
+			zap.Int64("segment", segmentID),
+			zap.Error(err))
 		return nil, err
 	}
 
-	// if the segment.EnableIndex == false, then load the segment immediately
-	if !enableIndex {
-		indexInfo.EnableIndex = false
-	} else {
-		indexInfo.BuildID = buildID
-		indexInfo.EnableIndex = true
-		err = broker.parseIndexInfo(ctx, segmentID, indexInfo)
+	infos, ok := resp.GetSegmentInfos()[segmentID]
+	if !ok {
+		log.Warn("segment not found",
+			zap.Int64("collection", collectionID),
+			zap.Int64("segment", segmentID))
+		return nil, fmt.Errorf("segment not found, collection: %d, segment: %d", collectionID, segmentID)
+	}
+
+	ret := make([]*querypb.VecFieldIndexInfo, 0, len(infos.IndexInfos))
+	for _, info := range infos.IndexInfos {
+		extraInfo, ok := infos.GetExtraIndexInfos()[info.IndexID]
```

Comment:
will it not panic when infos.getExtraIndexInfos() == nil?

Reply:
nope.

Reply:
```go
package main

import "fmt"

func main() {
    var m map[int]int = nil
    _, ok := m[1]
    fmt.Println(ok)
}
```